---
title: "R Notebook"
output: html_notebook
---
### load libraries, set wd
```{r}
library(tidyverse)
library(here)
here()

dr_here()
```
## GTDB Taxa
### update Daniel's sporulation table with updated GTDB taxa

```{r}


#### load data of known sporulation phenotype for all for all bacilliota (firmicutes) by gtdb family, based on Schwartz 2023

dan.tax <- read.csv(here("00_data/gtdb/original_metadata/gtdb_families_sporulation.csv" ))
dan.tax$gtdb_f<- gsub('f__', '', dan.tax$gtdb_f) 

### load recent gtdb tax table
gtdb.tax <- read.csv(here("00_data/gtdb/original_metadata/gtdb-taxonomy-table.csv" ))
gtdb.tax <- subset(gtdb.tax, gtdb.tax$Domain=="Bacteria")
colnames(gtdb.tax) <- paste0("newgtdb_", colnames(gtdb.tax))

## merge known spore data with updated gtdb tax info
tax <- merge(dan.tax, gtdb.tax, by.x="gtdb_f", by.y="newgtdb_Family", all.x=TRUE, all.y=TRUE)


### taxa from Schwartz 2023 that no longer match to any known families in GTDB, mostly spelling errors/mild inconsitencies. Needs manual fixing. 
weird <- subset(tax, is.na(tax$newgtdb_Domain))

### weird taxa removed for now.
tax <- subset(tax, !is.na(tax$newgtdb_Class))

tax <- unique(tax[,c(1,6:11)])


## make f_spor sporulation status as unknown if NA
tax$f_spor[is.na(tax$f_spor)] <- "unknown"

tax$f_spor <- ifelse(tax$newgtdb_Phylum!="Bacillota", FALSE, tax$f_spor)

### to pull out new bacilliota that have not had spor data updated:
#tax$f_spor <- ifelse(tax$newgtdb_Phylum=="Bacillota" & tax$f_spor=="unknown", "WEIRD", tax$f_spor)



write.csv(tax, here("00_data/gtdb/updated_fam_spor.csv"))
```

## INPHARED database

```{r}
### check inph for dereplication/ duplicates between refseq and genbank, ex: Bacillus phage Eldridge
#inph <- read.delim2(here(raw_dir, "1Jan2025_data.tsv"))
#inph$ignore <- "keep"
#inph <- inph[,c(1,28)]

## load inphared
inph<- read.delim2(here("00_data/inphared_db/original_metadata/14Apr2025_data_excluding_refseq.tsv"))
#inph.clean <- merge(inph, inph.noref, by="Accession", all.x=FALSE, all.y=TRUE)


replacements <- c(
  "Desulfobacterota_I" = "Desulfobacterota",
  "Bacteroidota_A"     = "Bacteroidota",
  "Bacillaceae_C"      = "Bacillaceae",
  "Bacillales_D"       = "Bacillales",
  "Bacillales_A"       = "Bacillales",
  "Bacillales_B"       = "Bacillales"
)


## add in sporulation status based on inphared host 
inph <- merge(inph, tax, by.x="Host", by.y="newgtdb_Genus", all.x=TRUE, all.y=FALSE)

## add in lifestyle predictions from phastyle
style <- read.delim2(here("00_data/inphared_db/inphared_phastyle_predictions.tsv"))

inph <- merge(inph, style, by.x="Accession", by.y="fasta_id", all.x=TRUE, all.y=FALSE)


# Apply all replacements to character columns
inph[] <- lapply(inph, function(x) {
  if (is.character(x)) {
    for (pattern in names(replacements)) {
      x <- gsub(pattern, replacements[[pattern]], x)
    }
  }
  x
})

## make f_spor sporulation status as unknown if NA
inph$f_spor[is.na(inph$f_spor)] <- "unknown"
inph$newgtdb_Phylum[is.na(inph$newgtdb_Phylum)] <- "unknown"

#### all phage with known sporulation status (T/F) of host, exclude any unknown phyla/sporulation status
inph.known <- subset(inph, inph$f_spor!="unknown")

### keep only phage with lifestyle predictions
inph.known <- subset(inph.known, !is.na(inph.known$predicted_label))


## create helpful labels
inph.known$sporulation <- "T"
inph.known$sporulation <- ifelse(inph.known$f_spor==TRUE, "Spor", "NonSpor")

inph.known$lifestyle <- "T"
inph.known$lifestyle <- ifelse(inph.known$predicted_label=="temperate", "Temp", "Lytic")

inph.known$host_phyla <- "T"
inph.known$host_phyla <- ifelse(inph.known$newgtdb_Phylum=="Bacillota", "Bacillota", "OtherPhyla")

inph.known <- unite(inph.known, "host_phage_spor", c("host_phyla", "lifestyle", "sporulation"), sep = "_", remove = FALSE, na.rm = FALSE)
inph.known <- unite(inph.known, "hostspec_phage_spor", c("newgtdb_Phylum", "lifestyle", "sporulation"), sep = "_", remove = FALSE, na.rm = FALSE)
inph.known <- unite(inph.known, "phage_type", c("lifestyle", "sporulation"), sep = "_", remove = FALSE, na.rm = FALSE)


## add in inphared dRep data (99.5% similarity clustered across 100% alignment)
### currently this is just test data w/ bacilliota
drep.w <- read.csv(here("00_data/inphared_db/inphared_bacill_drep_995-100_Wdb.csv"))
drep.w$ref<- "ref_seq"
drep.c <- read.csv(here("00_data/inphared_db/inphared_bacill_drep_995-100_Cdb.csv"))
drep.c <- drep.c[,c(1,8)]

drep <- merge(drep.w, drep.c, by="genome", all=TRUE)
drep$ref[is.na(drep$ref)] <- "dup"
drep$cluster <- drep$secondary_cluster
drep <- drep[,c(1,2,4)]

drep$genome<- gsub('.fna[^_]*$', '', drep$genome)



inph.known <- merge(inph.known, drep, by.x="Accession", by.y="genome", all=TRUE) 
inph.known$ref[is.na(inph.known$ref)] <- "not_run"
inph.known$cluster[is.na(inph.known$cluster)] <- "not_run"


unique(inph.known$newgtdb_Phylum)
```


```{r}


inph.phyla <- inph.known[,c(2,33)]
write.table(inph.phyla, here("00_data/inphared_db/14Apr2025_knownspore_phylalist.txt"), col.names = FALSE, row.names = FALSE, quote=FALSE, sep = "\t")

write.csv(inph.known, here("00_data/inphared_db/14Apr2025_knownsporestatus.csv"))
### only spore phages

inph.known <- read.csv(here("00_data/inphared_db/14Apr2025_knownsporestatus.csv"))
### create useful lists for splitting multifasta files

### list of all phage with KNOWN host sporulation status 
writeLines(as.character(inph.known$Accession), here("00_data/inphared_db/14Apr2025_KnownHostSporulation.txt"))


### list of all bacilliota sporulating (TRUE) hosts
writeLines(as.character(inph.known$Accession[inph.known$f_spor==TRUE]), here("00_data/inphared_db/14Apr2025_BacilliotataSporeTRUEphage.txt"))

## list of all bacilliota non-sporulating (FALSE) hosts
writeLines(as.character(inph.known$Accession[inph.known$f_spor==FALSE & inph.known$newgtdb_Phylum=="Bacillota"]), here("00_data/inphared_db/14Apr2025_BacilliotaSporeFALSEphage.txt"))

## list of all non-bacilliota and therefore non-sporulating hosts
writeLines(as.character(inph.known$Accession[inph.known$newgtdb_Phylum!="Bacillota"]), here("00_data/inphared_db/14Apr2025_NonBacilliotaphage.txt"))


### list of all bacilliota sporulating (TRUE) hosts + Lytic phage
writeLines(as.character(inph.known$Accession[inph.known$f_spor==TRUE & inph.known$predicted_label=="virulent"]), here("00_data/inphared_db/14Apr2025_BacilliotataSporeTRUELytic.txt"))

### list of all bacilliota NONsporulating (FALSE) hosts + Lytic phage
writeLines(as.character(inph.known$Accession[inph.known$f_spor==FALSE & inph.known$predicted_label=="virulent"]), here("00_data/inphared_db/14Apr2025_BacilliotataSporeFALSELytic.txt"))


### list of all bacilliota sporulating (TRUE) hosts + temp phage
writeLines(as.character(inph.known$Accession[inph.known$f_spor==TRUE & inph.known$predicted_label=="temperate"]), here("00_data/inphared_db/14Apr2025_BacilliotataSporeTRUETemp.txt"))

### list of all bacilliota NON sporulating (FALSE) hosts + temp phage
writeLines(as.character(inph.known$Accession[inph.known$f_spor==FALSE & inph.known$predicted_label=="temperate"]), here("00_data/inphared_db/14Apr2025_BacilliotataSporeFALSEemp.txt"))


## list of all non-bacilliota (and therefore non-sporulating hosts) and LYTIC
writeLines(as.character(inph.known$Accession[inph.known$newgtdb_Phylum!="Bacillota" & inph.known$predicted_label=="virulent"]), here("00_data/inphared_db/14Apr2025_NonBacilliotaLytic.txt"))

## list of all non-bacilliota (and therefore non-sporulating hosts) and temp
writeLines(as.character(inph.known$Accession[inph.known$newgtdb_Phylum!="Bacillota" & inph.known$predicted_label=="temperate"]), here("00_data/inphared_db/14Apr2025_NonBacilliotaTemp.txt"))


```


## GPD/GVD database, add sporulation status to gvd/gpd contigs -- UPDATED: 

```{r}

## load gtdb tax list with sporulation phenotype for bacilliota and non-bacilliota
tax <- read.csv(here("00_data/gtdb/updated_fam_spor.csv"),row.name=1)


search_terms <- unique(tax[,c(1)])

# Function to find first matching family in the host column from tax list and make a new column with just that family
extract_match <- function(text, terms) {
  match <- str_extract(text, str_c("(?i)", str_c(terms, collapse = "|"))) 
  ifelse(is.na(match), NA, match)
}


## load gpd and gvd datasets 

gpd <- read.delim2(here("00_data/gpd_gvd/original_metadata/GPD_metadata.tsv" )) 
gvd <- read.csv(here("00_data/gpd_gvd/original_metadata/gvd_all_hosts.csv" )) 

gpgv.style <- read.delim2(here("00_data/gpd_gvd/dan_phastyle_predictions.tsv"))




gpd.tax <- gpd[,c(1,7)]
gvd.tax <- gvd[,c(1,5)]
## can use this to match to daniel's GTDB spore/nonspore and phagescope output

colnames(gpd.tax) <- c("ID", "GTDB_tax")
colnames(gvd.tax)<- c("ID", "GTDB_tax")

dan.tax <- rbind(gpd.tax, gvd.tax)

dan.tax <- unique(dan.tax)


dan.tax <- subset(dan.tax, dan.tax$Completeness=="Complete"| dan.tax$Completeness=="High-quality" | dan.tax$Completeness=="Medium-quality")

```

```{r}

dan.tax <- merge(dan.tax, gpgv.style, by.x="ID", by.y="fasta_id", all.x=TRUE, all.y=TRUE)

dan.tax$GTDB_tax[is.na(dan.tax$GTDB_tax)] <- "Host Not Assigned or not a prokaryotic virus"

dan.tax2 <- dan.tax %>%
  mutate(matched_word = sapply(GTDB_tax, extract_match, terms = search_terms))



dan.tax2$matched_word[is.na(dan.tax2$matched_word)] <- "Unknown"


##remove genus to prevent duplicates
tax_f <- tax[,c(1:6)]
tax_f <- unique(tax_f)
dan.tax2 <- unique(dan.tax2)

dan.tax3 <- merge(tax_f, dan.tax2, by.x="gtdb_f", by.y="matched_word", all.x =FALSE, all.y=TRUE)

dan.tax3$predicted_label[is.na(dan.tax3$predicted_label)] <- "missing"

missingpred <- subset(dan.tax3, dan.tax3$predicted_label=="missing")
#dan.tax4 <- unique(dan.tax3)

#write.csv(dan.tax3, here("00_data/gpd_gvd/gpd_gvd_sporstatus_checkv.csv"))

```

### Group GPD/GVD contigs by host/lifestyle
```{r}
library(here)
gpdgvd <- read.csv(here("00_data/gpd_gvd/gpd_gvd_sporstatus.csv"), row.names=1)

ps.1 <- read.delim2(here("00_data/gpd_gvd/original_metadata/phagescope_gpd_phage_meta_data.tsv"))
ps.2 <- read.delim2(here("00_data/gpd_gvd/original_metadata/phagescope_gvd_phage_meta_data.tsv"))

ps <- rbind(ps.1, ps.2)

ps.checkv <- ps[,c(1:6)]

gpdgvd2 <- merge(gpdgvd, ps.checkv, by.x="ID", by.y="Phage_ID", all.x=TRUE, all.y=FALSE)
```


```{r}
### create useful lists for splitting multifasta files
gpdgvd <- read.csv(here("00_data/gpd_gvd/gpd_gvd_sporstatus_checkv.csv"))

writeLines(as.character(gpdgvd$ID), here("00_data/gpd_gvd/GPDGVD_Phage_qual.txt"))

### list of all bacilliota sporulating (TRUE) hosts
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==TRUE]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeTRUEphage_qual.txt"))

## list of all bacilliota non-sporulating (FALSE) hosts
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==FALSE & gpdgvd$newgtdb_Phylum=="Bacillota"]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeFALSEphage_qual.txt"))

### list of all bacilliota sporulating (TRUE) hosts + Lytic phage
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==TRUE & gpdgvd$predicted_label=="virulent"]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeTRUELytic_qual.txt"))

### list of all bacilliota non sporulating (FALSE) hosts + Lytic phage
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==FALSE & gpdgvd$predicted_label=="virulent"]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeFALSELytic_qual.txt"))

### list of all bacilliota sporulating (TRUE) hosts + Temp phage
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==TRUE & gpdgvd$predicted_label=="temperate"]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeTRUETemperate_qual.txt"))

### list of all bacilliota sporulating (FALSE) hosts + Temp phage
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==TRUE & gpdgvd$predicted_label=="temperate"]), here("00_data/gpd_gvd/GPDGVD_BacilliotaSporeFALSETemperate_qual.txt"))

## list of all bacilliota non-sporulating (FALSE) hosts
writeLines(as.character(gpdgvd$ID[gpdgvd$f_spor==FALSE]), here("00_data/gpd_gvd/GPDGVD_NonBacilliotaPhage_qual.txt"))

```


